### DISCUSSION TITLE
Log Activity and Write Telemetry in Constructor

### DISCUSSION THREAD

**Discussion by MichaelStarks** [nasa/fprime#1412] [Q&A] [answered]
I have a State Machine object and would like to log the state telemetry and activity in the constructor. When I move the `this->log_ACTIVITY_HI_STATE_CHANGE();` and `this->tlmWrite_STATE();` to the constructor I fail the unit test that I have written. When I remove the portion of the unit test that checks the start-up state it passes, from this I assume the activity and telemetry are either not being logged or are being cleared. How should I setup setting the start-up state so that the information is logged?

Thank you

**Comment by keck-in-space** [comment]
I wanted to follow up on this question -- is there any way to log in the constructor of a component?

**Comment by SterlingPeet** [comment, marked as answer]
F Prime's logging and telemetry handling is not guaranteed to be initialized at the time of component construction.  Because they are themselves components, this would be a race condition.  For those mechanisms to work, the topology ports need to be connected; a process that starts _after_ all of the component instances have been constructed and initialized.  BTW; that also means you won't have access to spacecraft time via ``getTime()`` either.

So, if you want to log something during construction for debugging and testing purposes, you could go around the built-in tooling and just `printf()` that information.  If you need that information to go through a flight or flight-like downlink to a GDS, you will have to cache that information somewhere like member variables and downlink it at a later time.
